concepts/0519-goal-codes/README.md

```diff
@@ -0,0 +1,149 @@
+- Authors: [Daniel Hardman](daniel.hardman@gmail.com)
+- Status: [PROPOSED](/README.md#proposed)
+- Since: 2020-07-24
+- Status Note: surfaces thinking first used in [0434 OOB Protocol](../../features/0434-outofband/README.md) and in [Co-protocols](https://docs.google.com/presentation/d/17hk6QqLW5M9E4TBPZwXIUBu9eEinMNXReceyZTF4LpA/edit)
+- Start Date: 2020-01-01
+- Tags: concept
+
+## Summary
+
+Explain how different parties in an SSI ecosystem can communicate about their intentions in a way that is understandable by humans and by automated software.
+
+## Motivation
+
+[Agents](../0004-agents/README.md) exist to achieve the intents of their owners. Those intents largely unfold through [protocols](../0003-protocols/README.md). Sometimes intelligent action in these protocols depends on a party declaring their intent. We need a standard way to do that.
+
+## Tutorial
+
+Our early learnings in SSI focused on VC-based proving with a very loose, casual approach to context. We did demos where Alice connects with a potential employer, Acme Corp -- and we assumed that each of the interacting parties had a shared understanding of one another's needs and purposes.
+
+But in a mature SSI ecosystem, where unknown agents can contact one another for arbitrary reasons, this context is not always easy to deduce. Acme Corp's agent may support many different protocols, and Alice may interact with Acme in the capacity of customer or potential employee or vendor. Although we have [feature discovery](../../features/0031-discover-features/README.md) to learn what's possible, and we have [machine-readable governance frameworks](../0430-machine-readable-governance-frameworks/README.md) to tell us what rules might apply in a given context, we haven't had a way to establish the context in the first place. When Alice contacts Acme, a context is needed before a governance framework is selectable, and before we know which features are desirable.
+
+The key ingredient in context is __intent__. If Alice says to Acme, "I'd like to connect,", Acme wants to be able to trigger different behavior depending on whether Alice's intent is to be a customer, apply for a job, or audit Acme's taxes. This is the purpose of a goal code.
+
+### The goal code datatype
+
+To express intent, this RFC formally introduces the goal code datatype. When a field in a [DIDComm](../0005-didcomm/README.md) message contains a goal code, its semantics and format match the description given here. (Goal codes are often declared via the `~thread` decorator, but may also appear in ordinary message fields. See the [Scope section](#scope) below. Convention is to name this field "goal_code" where possible; however, this is only a convention, and individual protocols may adapt to it however they wish.)
+
+>TODO: should we make a decorator out of this, so protocols don't have to declare it, and so any message can have a goal code? Or should we just let protocols declare a field in whatever message makes sense?
```

Comment:
We don't currently have the concept of "inheritable *fields*" so a decorator is the best we can do that can be applied across all protocols.

And since we would add a decorator anyway, I suggest we add it to `~thread` just like you did in your threading example above.